ui/user/src/lib/components/mcp/McpResources.svelte

```diff
@@ -0,0 +1,451 @@
+	};
+
+	let { project, threadID = $bindable(), currentThreadFiles = $bindable() }: Props = $props();
+	let dialog = $state<HTMLDialogElement>();
+	let loading = $state(false);
+	let projectResources = $state<ServerResources[]>([]);
+	let targetResourceUri = $state('');
+	let searchQuery = $state('');
+	let searchInput = $state<HTMLInputElement>();
+	let loadingFiles = $state(false);
+
+	const fileExtensions: Record<string, string> = {
+		'text/plain': 'txt',
+		'text/markdown': 'md',
+		'application/javascript': 'js',
+		'application/typescript': 'ts',
+		'application/octet-stream': 'bin'
+	};
+
+	// Sort MCP servers lexicographically by name + mcpID
+	let projectMcps = $derived(
+		[...getProjectMCPs().items].sort((a, b) => getMcpFQN(a).localeCompare(getMcpFQN(b)))
+	);
+
+	// Fuzzy search function
+	function fuzzyMatch(query: string, text: string): boolean {
+		if (!query) return true;
+		const searchLower = query.toLowerCase();
+		const textLower = text.toLowerCase();
+		return textLower.includes(searchLower);
+	}
+
+	// Filtered resource sets based on search query
+	let filteredResources = $derived(
+		projectResources
+			.map((serverResources) => ({
+				...serverResources,
+				resources: serverResources.resources.filter(
+					(resource) =>
+						fuzzyMatch(searchQuery, resource.name) ||
+						fuzzyMatch(searchQuery, getMcpFQN(serverResources.mcp))
+				)
+			}))
+			.filter(
+				(serverResources) =>
+					fuzzyMatch(searchQuery, getMcpFQN(serverResources.mcp)) ||
+					serverResources.resources.length > 0
+			)
+	);
+
+	function fetchProjectResources() {
+		loading = true;
+		projectResources = [];
+		for (const mcp of projectMcps) {
+			if (!mcp.authenticated) {
+				projectResources.push({
+					mcp,
+					resources: []
+				});
+				continue;
+			}
+
+			ChatService.listProjectMcpServerResources(project.assistantID, project.id, mcp.id)
+				.then((resources) => {
+					if (resources.length < 1) {
+						return;
+					}
+
+					projectResources.push({
+						mcp,
+						resources
+					});
+				})
+				.catch((error) => {
+					// 424 means resources not supported
+					if (!error.message.includes('424')) {
+						console.error('Failed to load resources for MCP server:', mcp.id, error);
+					}
+				});
+		}
+		loading = false;
+	}
+
+	async function loadThreadFiles() {
+		if (!threadID) return;
+		loadingFiles = true;
+		try {
+			const files = await ChatService.listFiles(project.assistantID, project.id, { threadID });
+			currentThreadFiles = files.items;
+		} catch (err) {
+			errors.append(`Failed to load thread files: ${err}`);
+			currentThreadFiles = [];
+		}
+		loadingFiles = false;
+	}
+
+	function getMcpFQN(mcp: ProjectMCP) {
+		return (mcp.name ?? '') + '-' + mcp.id;
+	}
+
+	function getFilename(mcp: ProjectMCP, resourceName: string, mimeType: string) {
+		const extension = fileExtensions[mimeType] ?? mimeType.split('/')?.[1] ?? 'txt';
+		const filename = `obot-${getMcpFQN(mcp)}-resource-${resourceName}.${extension}`;
```

Comment:
Not crazy about the naming, but we need something unique per MCP-server/instance to key off so we can discover existing resource files in the workspace.

More of a long-term idea, but it would be interesting if we had a "directory" in the workspace reserved for these types of files, and they were immutable to the `workspace-files` tool and stamped with the hash of the content so we can tell when they've changed compared to what the MCP server is returning. We could then offer an "update file" option that gets show when drift is detected.